app/services/inventory_check_service.rb

```diff
@@ -0,0 +1,61 @@
+class InventoryCheckService
+  attr_reader :error, :alert
+
+  def initialize(distribution)
+    @distribution = distribution
+    @alert = nil
+    @error = nil
+  end
+
+  def call
+    unless items_below_minimum_quantity.empty?
+      set_error
+    end
+
+    unless deduplicate_items_below_recommended_quantity.empty?
+      set_alert
+    end
+
+    self
+  end
+
+  def set_error
+    @error = "The following items have fallen below the minimum " \
+      "on hand quantity: #{items_below_minimum_quantity.map(&:name).sort.join(", ")}"
+  end
+
+  def set_alert
+    @alert = "The following items have fallen below the recommended " \
+      "on hand quantity: #{deduplicate_items_below_recommended_quantity.map(&:name).sort.join(", ")}"
+  end
+
+  def items_below_minimum_quantity
+    unless @items_below_minimum_quantity
+      item_ids = @distribution.line_items.select do |line_item|
+        inventory_item = line_item.item.inventory_item_at(@distribution.storage_location.id)
+        inventory_item.lower_than_on_hand_minimum_quantity?
+      end.map(&:item_id)
+
+      @items_below_minimum_quantity = Item.find(item_ids)
+    end
+
+    @items_below_minimum_quantity
+  end
```

Comment:
@albertchae This is a bit wordier but should avoid the N+1 query pattern. Let me know if you think there's a nicer way to write this.

Reply:
Hmm, an alternate way is to do `@distribution.line_items.includes(:item).select` to start the block so the `item` association is preloaded, but I don't know if that could potentially load more than is necessary (since you are filtering out inventory items that would fail the `lower_than_on_hand_minimum_quantity?`). 

At this point I think either form is fine, so your call. If you go ahead with the map to `item_id` and explicit `Item.find`, I would at least leave a comment saying this is to avoid an N+1.

Reply:
Added comments!

Reply:
Thanks @egiurleo